Sources/GRPC/ClientCalls/Call.swift

```diff
@@ -162,6 +177,84 @@ extension Call {
   }
 }
 
+extension Call {
+  internal func compress(_ compression: Compression) -> Bool {
+    return compression.isEnabled(callDefault: self.isCompressionEnabled)
+  }
+
+  internal func sendMessages<Messages>(
+    _ messages: Messages,
+    compression: Compression,
+    promise: EventLoopPromise<Void>?
+  ) where Messages: Sequence, Messages.Element == Request {
+    if self.eventLoop.inEventLoop {
+      if let promise = promise {
+        self._sendMessages(messages, compression: compression, promise: promise)
+      } else {
+        self._sendMessages(messages, compression: compression)
+      }
+    } else {
+      self.eventLoop.execute {
+        if let promise = promise {
+          self._sendMessages(messages, compression: compression, promise: promise)
+        } else {
+          self._sendMessages(messages, compression: compression)
+        }
+      }
+    }
+  }
+
+  // Provide a few convenience methods we need from the wrapped call objects.
+  private func _sendMessages<Messages>(
+    _ messages: Messages,
+    compression: Compression
+  ) where Messages: Sequence, Messages.Element == Request {
+    self.eventLoop.assertInEventLoop()
+    let compress = self.compress(compression)
+
+    var iterator = messages.makeIterator()
+    var maybeNext = iterator.next()
+    while let current = maybeNext {
+      let next = iterator.next()
+      // If there's no next message, then we'll flush.
+      let flush = next == nil
+      self._send(.message(current, .init(compress: compress, flush: flush)), promise: nil)
+      maybeNext = next
+    }
+  }
+
+  private func _sendMessages<Messages>(
+    _ messages: Messages,
+    compression: Compression,
+    promise: EventLoopPromise<Void>
+  ) where Messages: Sequence, Messages.Element == Request {
+    self.eventLoop.assertInEventLoop()
+    let compress = self.compress(compression)
+
+    var futures: [EventLoopFuture<Void>] = []
+    futures.reserveCapacity(messages.underestimatedCount)
+
+    var iterator = messages.makeIterator()
+    var maybeNext = iterator.next()
+    while let current = maybeNext {
+      let next = iterator.next()
+      // If there's no next message, then we'll flush.
+      let flush = next == nil
+
+      // Make a promise, store the future.
+      let promise = self.eventLoop.makePromise(of: Void.self)
```

Comment:
Do we need a promise on every message? It seems sufficient to attach one to the _last_ message, doesn't it? For streaming protocols like this you'd expect all these writes to fail with the same error.

Reply:
Hmm, I guess not? I suppose it's technically possible for one write to fail if a handler does something weirdo but practically speaking that's highly unlikely to happen. Let's save some allocations.